### DISCUSSION TITLE
dispatch an action in getServerSideProps with redux-toolkit

### DISCUSSION THREAD

**Discussion by sabermazoji** [vercel/next.js#20122] [Help]
I wanna access the store and dispatch an action in  getServerSideProps or getInitialProps,
I use this example https://github.com/vercel/next.js/tree/canary/examples/with-redux-toolkit.

I'm trying to add my store to the page like this but my client-side store doesn't update!

store.js 
```
import { configureStore } from '@reduxjs/toolkit';
import authReducer from './slices/authSlice';
import buySellSlice from './slices/buysellSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    buysell: buySellSlice,
  },
  devTools: true,
});

```
About.js (nextjs page)

```
import store from '../../store';
const About=()=>{
return(<>about us</>);
}
export async function getServerSideProps(props) {
  await store.dispatch(getCoinByTitle('BTC_IRR'));
  return { props: { data: '' } };
}
export default About;
```

**Comment by oMatej** [comment]
You should not dispatch any redux action inside `getServerSideProps`. You should return props that will be used to initialize the `redux` store with proper values.

Something like:

```
const Page = ({ initialState }) => {
  const store = createStore(initialState);

  return (
    <Provider store={store}> ... </Provider>
  );
}

export const getServerSideProps = () => {
  const initialState = /* get initial data for redux store */;

  return {
    props: { initialState }
  }
}

export default Page;
```

Since you have your `Provider` in `_app.js`, then you can try to do something like:

```
const MyApp = ({ Component, pageProps }) => {
  const { initialState, ...restPageProps } = pageProps;

  const store = createStore(initialState);

  return (
    <Provider store={store}>
      <Component {...restPageProps} />
    </Provider>
  )
}
```

You would also have to change a bit the implementation of `store`, so that it will not be recreated on every page rerender. Maybe to something like this:

```
// store.js

let store;

const createStore = (preloadedState) => {
  if (typeof window === 'undefined') {
    return configureStore({
      reducer: { ... },
      preloadedState,
    });
  }

  if (!store) {
    store = configureStore({
      reducer: { ... },
      preloadedState,
    });
  }

  return store;
};

export default createStore;
```

**Reply by sabermazoji** [reply]
so for data fetching, I should send HTTP request in getServerSideProps and return data with props then dispatch or update my store in the component? 
like this?
```
import { useDispatch } from 'react-redux';
const About=({data})=>{
const dispatch = useDispatch();
dispatch(getList(data));

return(<>about us</>);
}
export async function getServerSideProps(props) {
    const res = await fetch('..')
  return { props: { data:res.data} };
}
export default About;
```

**Reply by pjtf93** [reply]
Hi, did you found a solution for this?

**Reply by mkubdev** [reply]
@sabermazoji Look ok for me!

**Reply by Abusayid693** [reply]
Just wondering if this is a good approach....?

**Reply by brayanduranvelasquez** [reply]
I have the same question. If it is a good approach.

**Reply by selimdev00** [reply]
It seems to be a good solution, because even with server components (Next.js 13 and higher) you cannot use basic functions of redux inside server components. The only reasons we would fetch data from server in the SSR of Next.js is to boost SEO and use caching & revalidating features of Next.js.
